rust/gelato/src/fwd_req_op.rs

```diff
@@ -0,0 +1,152 @@
+use crate::chains::Chain;
+use crate::err::GelatoError;
+use crate::fwd_req_call::ForwardRequestCall;
+use crate::task_status_call::{
+    TaskStatus, TaskStatusCall, TaskStatusCallArgs,
+};
+use ethers::signers::Signer;
+use ethers::types::{Address, Bytes, U256};
+use std::sync::Arc;
+use tokio::time::Duration;
+use tracing::{info, instrument};
+
+extern crate num_derive;
+
+#[derive(Debug, Clone)]
+pub struct ForwardRequestOpArgs {
+    pub chain_id: Chain,
+    pub target: Address,
+    pub data: Bytes,
+    pub fee_token: Address,
+    pub payment_type: PaymentType,
+    pub max_fee: U256,
+    pub gas: U256,
+    pub sponsor: Address,
+    pub sponsor_chain_id: Chain,
+    pub nonce: U256,
+    pub enforce_sponsor_nonce: bool,
+    pub enforce_sponsor_nonce_ordering: bool,
+}
+
+#[derive(Debug, Clone)]
+pub struct ForwardRequestOpResult {
+    // TODO: lots more here than just task_id,
+// including gas charged, block number, etc
+// etc etc etc.
+}
+
+#[derive(Debug, Clone)]
+pub struct ForwardRequestOptions {
+    pub poll_interval: Duration,
+    pub retry_submit_interval: Duration,
+}
+
+// TODO(webbhorn): Support cancel() on an in-flight
+// ForwardRequestOp from some task other than the one
+// currently inside of op.run(), e.g. by using a
+// tokio_util::sync::CancellationToken
+// (https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html)
+// that is private to the Op struct and signalled via a
+// call to op.cancel().
+//
+// TODO(webbhorn): ... And/Or, support injection of a
+// RetryPolicy closure in construction of the Op that
+// we can specifically use to query the shared Inbox
+// state to e.g. find out if a message has already
+// been delivered (perhaps by some other Relayer), and
+// we should stop retrying and return from run().
+//
+// TODO(webbhorn): Maybe take deps on traits
+// instead of concrete types, e.g. for 'http'.
+#[derive(Debug, Clone)]
+pub struct ForwardRequestOp<S> {
+    pub args: ForwardRequestOpArgs,
+    pub opts: ForwardRequestOptions,
+    pub signer: S,
+    pub http: Arc<reqwest::Client>,
+    // TODO: tracer?
+    // TODO: interface to DB?
+    // TODO: prometheus registry?
+}
+
+impl<S: Signer> ForwardRequestOp<S> {
+    #[instrument]
+    pub async fn run(
+        &self,
+    ) -> Result<ForwardRequestOpResult, GelatoError> {
+        // TODO(webbhorn): handle signing error. Presumably for AWS
+        // some are retryable, others not?
+        let sig = self.signer.sign_typed_data(&self.args).await.unwrap();
+        loop {
+            let fwd_req_call = ForwardRequestCall {
+                http: Arc::clone(&self.http),
```

Comment:
intended to come back and plumb these all by ref rather than cloning everything, was just done for expediency at first

Reply:
Resolving this since I've punted fwd_req_op.rs until a future PR.